### ISSUE TITLE
c.tti metric calculation

### ISSUE THREAD

**Issue by judwhite** [akamai/boomerang#245]
**tl;dr**: `c.tti` reported a value of over 2 minutes when it appears it should have been closer to 860ms. I suspect having a Service Worker was a contributing factor.

Moved from another issue to prevent issue hijacking.

I'm re-evaluating `c.tti`'s meaning, while trying to make sense of these logs.

The interesting part is `c.tti=126350` (two minute TTI!) in the second beacon. The first beacon has `c.tti.vr=858` while missing `c.tti`, and both beacons have `t_done=860`.

The time from `/` to the first beacon is two minutes. Interestingly the last request to `/` is from a Service Worker and comes directly _after_ the two beacons in the logs. The Service Worker does attempt to cache `/` on install.

This is the user's first visit based on IP. They're using a recent version of Chrome on an Android device released within the last 2 years, and using a mobile network within the United States.

NGINX logs. I don't log requests to images, css, or js, so can't provide any clues there.

```
###.###.##.### - - [11/Feb/2019:14:54:27 -0600] "GET /?gclid=sOmEGcLiDdAtA HTTP/2.0" 200 8899 "https://www.google.com/" "Mozilla/5.0 ..."
###.###.##.### - - [11/Feb/2019:14:54:28 -0600] "GET /site.webmanifest HTTP/2.0" 200 1458 "https://myorigin.com/?gclid=sOmEGcLiDdAtA" "Mozilla/5.0 ..."
###.###.##.### - - [11/Feb/2019:14:56:41 -0600] "POST /api/v1/beacon/boomr HTTP/2.0" 204 0 "https://myorigin.com/?gclid=sOmEGcLiDdAtA" "Mozilla/5.0 ..."
###.###.##.### - - [11/Feb/2019:14:56:57 -0600] "POST /api/v1/beacon/boomr HTTP/2.0" 204 0 "https://myorigin.com/?gclid=sOmEGcLiDdAtA" "Mozilla/5.0 ..."
###.###.##.### - - [11/Feb/2019:14:56:57 -0600] "GET / HTTP/2.0" 200 8899 "https://myorigin.com/service-worker.js" "Mozilla/5.0 ..."
```

BOOMR Configuration:

```js
BOOMR.init({
    autorun: true,
    beacon_type: "POST",
    beacon_url: '/api/v1/beacon/boomr',
    instrument_xhr: true,
    RT: {
        enabled: true,
        cookie: true,
        strict_referrer: true,
    },
    Errors: {
        enabled: true,
        sendAfterOnload: true,
        monitorRejections: true,
    },
    Continuity: {
        enabled: true,
        afterOnload: true,
        waitAfterOnload: 1000,
        monitorStats: true,
        monitorInteractions: true,
        sendTimeline: false,
        isSpa: true
    },
    AutoXHR: {
        enabled: true,
        monitorFetch: true,
        alwaysSendXhr: false,
    },
    log: function () {},
    /*History: {
        enabled: true,
        auto: true
    },
    ResourceTiming: {
        enabled: true,
        clearOnBeacon: true,
    }*/
});
```

Backend Logs, in the order written to disk.

Request 1 (1/4):

```
time="2019-02-11T14:54:27-06:00" 
	level=info 
	msg=OK 
	bytes_sent=8899 
	event=index 
	http_status=200 
	method=GET 
	uri="/?gclid=..."
```

Beacon 1 (2/4):

```
time="2019-02-11T14:56:41-06:00" 
	level=info
	msg="No Content" 
	boomr.c.e=1549918465007 
	boomr.c.f=2 
	boomr.c.f.d=2490 
	boomr.c.f.l=2 
	boomr.c.f.m=1 
	boomr.c.f.s=1549918465858 
	boomr.c.ttfi=1053 
	boomr.c.tti.m=lt 
	boomr.n=1 
	boomr.nt_con_end=1549918465316 
	boomr.nt_con_st=1549918465316 
	boomr.nt_dns_end=1549918465316 
	boomr.nt_dns_st=1549918465316 
	boomr.nt_domcomp=1549918465866 
	boomr.nt_domcontloaded_end=1549918465865 
	boomr.nt_domcontloaded_st=1549918465864 
	boomr.nt_domint=1549918465864 
	boomr.nt_domloading=1549918465421 
	boomr.nt_fet_st=1549918465316 
	boomr.nt_load_end=1549918465866 
	boomr.nt_load_st=1549918465866 
	boomr.nt_nav_st=1549918465007 
	boomr.nt_nav_type=2 
	boomr.nt_req_st=1549918465316 
	boomr.nt_res_end=1549918465466 
	boomr.nt_res_st=1549918465410 
	boomr.nt_worker_start=1549918465023 
	boomr.pid=0gvclolm 
	boomr.r="https://www.google.com/" 
	boomr.rt.bmr="518,293,245,6,4,,4,4,4" 
	boomr.rt.bstart=1549918465835 
	boomr.rt.end=1549918465867 
	boomr.rt.si=tycjfgmuarb-pms4qp 
	boomr.rt.sl=1 
	boomr.rt.ss=1549918465007 
	boomr.rt.start=navigation 
	boomr.rt.tstart=1549918465007 
	boomr.rt.tt=860 
	boomr.t_done=860 
	boomr.t_other="[]boomr.NameValue{boomr.NameValue{Name:\"t_domloaded\", Value:858}, boomr.NameValue{Name:\"boomerang\", Value:10}, boomr.NameValue{Name:\"boomr_fb\", Value:828}}" 
	boomr.t_page=456 
	boomr.t_resp=404 
	boomr.ua.plt="Linux armv8l" 
	boomr.ua.vnd="Google Inc." 
	boomr.v=1.0.0 
	boomr.vis.lh=2282 
	boomr.vis.st=hidden 
	boomr_unk.c.l="4t9,s1" 
	boomr_unk.c.lt="[{\"s\":\"mx\",\"d\":\"1e\",\"n\":1}]" 
	boomr_unk.c.lt.n=1 
	boomr_unk.c.lt.tt=50.19999999785796 
	boomr_unk.c.tti.vr=858 
	bytes_sent=0 
	event=api-beacon-boomr
	http_status=204 
	method=POST 
	uri=/api/v1/beacon/boomr
```

Beacon 2 (3/4):

```
time="2019-02-11T14:56:57-06:00" 
	level=info
	msg="No Content" 
	boomr.c.e=1549918465007 
	boomr.c.f=3 
	boomr.c.f.d=130018 
	boomr.c.f.l=1 
	boomr.c.f.m=4 
	boomr.c.f.s=1549918468354 
	boomr.c.lb=1549918468350 
	boomr.c.ttfi=1053 
	boomr.c.tti=126350         ### <-- I'd be fired :)
	boomr.c.tti.m=lt 
	boomr.n=2 
	boomr.pid=0gvclolm 
	boomr.r="https://www.google.com/" 
	boomr.rt.bstart=1549918465835 
	boomr.rt.end=1549918465867 
	boomr.rt.quit=true 
	boomr.rt.si=tycjfgmuarb-pms4qp 
	boomr.rt.sl=1 
	boomr.rt.ss=1549918465007 
	boomr.rt.tstart=1549918465007 
	boomr.rt.tt=860 
	boomr.t_done=860 
	boomr.ua.plt="Linux armv8l" 
	boomr.ua.vnd="Google Inc." 
	boomr.v=1.0.0 
	boomr.vis.lh=132310 
	boomr.vis.lv=6939 
	boomr.vis.st=visible 
	boomr_unk.c.l="42pjs,s0" 
	boomr_unk.c.tti.vr=858 
	bytes_sent=0 
	event=api-beacon-boomr 
	http_status=204
	method=POST
	uri=/api/v1/beacon/boomr
```

Request 2 (4/4):

```
time="2019-02-11T14:56:57-06:00" 
	level=info 
	msg=OK 
	bytes_sent=8899 
	event=index 
	http_status=200 
	method=GET 
	uri=/
```

---

For completeness, I'm using 6edb5cf50511123d325989a1eff9982962eeac3e.

plugins.json:

```js
{
    "plugins": [
        "plugins/rt.js",
        "plugins/navtiming.js",

        "plugins/spa.js",
        "plugins/auto-xhr.js",
        "plugins/history.js",

        "plugins/continuity.js",

        "plugins/errors.js",

        "plugins/usertiming.js",

        "plugins/compression.js",
        "plugins/md5.js"
    ]
}
```

**Comment by judwhite** [comment]
@bluesmoon @nicjansma The information is there. I don't care what you do with it, so I'm closing this issue as I have no plans to trust `c.tti`.

**Comment by nicjansma** [comment]
Hi @judwhite!  Thanks for opening this issue.

Are you seeing TTI being reported this high in many cases, or is this an outlier?

With all RUM data, we find it's best to not focus on the outliers.  There are many reasons metrics/timers can show extreme values:

* Old browsers
* Browser bugs
* Corruption or modification on the wire
* Library incompatibility with old browsers
* Library bugs
* etc

We see all of the above, all of the time, in our RUM data.  Most of our customers are generally looking at medians or Nth percentiles, which can help take focus away from the cases that look extreme.

In this case, I have a few ideas of what happened.

Beacon 1 is the main Page Load beacon, though it was a Back-Forward navigation:

```
	boomr.c.ttfi=1053 // Time to First Interaction (when the user first clicked)
	boomr.c.tti.m=lt  // TTI calculation method (LongTasks supported)
...
	boomr.nt_load_end=1549918465866 // Page Load complete (859ms)
...
	boomr.nt_nav_st=1549918465007 // Page Load start
...
	boomr.nt_nav_type=2 // Back-Forward nav
...
	boomr.rt.sl=1 // Session Length of 1, but since it's a Back-Forward, my guess is their prev session timed out
...
	boomr.rt.tt=860 // confirmation of total Page Load time of 860ms
	boomr.t_done=860  // confirmation of total Page Load time of 860ms
...
	boomr.vis.lh=2282 // last time we went hidden was at 2282ms
	boomr.vis.st=hidden // the beacon was sent while hidden
...
	boomr_unk.c.tti.vr=858 // Visually Ready at 858ms (right before Page Load)
```

The important parameter above is that the beacon was sent while hidden (`vis.st=hidden`), so it's possible that much of the page was loading while hidden.

We don't see TTI (`c.tti`) on this Page Load beacon.  My guess is that this is because of the short `waitAfterOnload: 1000` (wait up to 1 second), and, probably because the page was hidden while loading.  Background tabs may confuse the TTI measurement.

The important parameter in Beacon 2 is that it's an "unload" beacon:

```
boomr.c.tti=126350 // Finally a TTI
...
boomr.rt.quit=true // Unload beacon
...
// visibility data
boomr.vis.lh=132310  // Last Hidden timestamp
boomr.vis.lv=6939  // Last Visible timestamp
boomr.vis.st=visible // Current state
```

Unload beacons are (attempted to be) sent as the user is navigating away.  We primarily use these beacons for Session Duration measurements (how long they were on the page).  The metrics/timers/etc will often be repeated on the Unload beacon -- I wouldn't use any data from the Unload beacon, as they would then be double-counted.

In this case, we do see `c.tti` show up only on the Unload beacon, and not the Page Load beacon, with this extreme value.  My guess is this is related to the visibility state.  TTI couldn't be calculated at the Page Load beacon (it didn't happen yet).  The user probably re-tabbed to the page and/or then navigated it away, which triggered the Unload beacon.  This unload beacon may have finally been able to calculate TTI, but knowing it's bogus due to the Visibility State, we should ignore it.

So our recommendations would be:

* Take Page Loads of `vis.st=hidden` with a grain of salt, or, bucket them differently
* Know there will always be outliers in your data and focus on the median or Nth percentile
* Ignore everything in the Unload beacon except for Session Duration

There may be bugs in the TTI measurement if loaded while Hidden.  We'll look into this.
